content.en/tools/tiros.md

```diff
@@ -0,0 +1,230 @@
+
+```go
+browser := rod.New().Context(ctx).ControlURL("ws://localhost:3000")) // default CDP chrome port
+
+browser.Connect()
+defer browser.Close()
+
+var metricsStr string
+rod.Try(func() {
+    browser = browser.Context(c.Context).MustIncognito()
+    browser.MustSetCookies() // clears cookies if parameter is empty
+    
+    page := browser.MustPage() // Get a handle of a new page in our incognito browser
+    
+    page.MustEvalOnNewDocument(jsOnNewDocument) // clears the cache by running `localStorage.clear()`
+    
+    // disable caching in general
+    proto.NetworkSetCacheDisabled{CacheDisabled: true}.Call(page) // make sure we're not hitting the cache
+
+
+    // finally navigate to url and fail out of rod.Try by panicking
+    page.Timeout(websiteRequestTimeout).Navigate(url)
+    page.Timeout(websiteRequestTimeout).WaitLoad() // waits for onload event
+    page.Timeout(websiteRequestTimeout).WaitIdle(time.Minute)
+
+    page.MustEval(wrapInFn(jsTTIPolyfill)) // add TTI polyfill
+    page.MustEval(wrapInFn(jsWebVitalsIIFE)) // add web-vitals
+
+    // finally actually measure the stuff
+    metricsStr = page.MustEval(jsMeasurement).Str()
+    
+    page.MustClose()
+})
+// parse metricsStr
+```
+
+`jsOnNewDocument` contains javascript that gets executed on a new page before anything happens. We're subscribing to performance events which is necessary for TTI polyfill and we're clearing the local storage. This is the code ([link to source](https://github.com/dennis-tra/tiros/blob/main/js/onNewDocument.js)):
+
+```javascript
+// From https://github.com/GoogleChromeLabs/tti-polyfill#usage
+!function(){if('PerformanceLongTaskTiming' in window){var g=window.__tti={e:[]};
+    g.o=new PerformanceObserver(function(l){g.e=g.e.concat(l.getEntries())});
+    g.o.observe({entryTypes:['longtask']})}}();
+
+localStorage.clear();
+```
+
+Then, after the website has loaded we are adding a [TTI polyfill](https://github.com/dennis-tra/tiros/blob/main/js/tti-polyfill.js) and [web-vitals](https://github.com/dennis-tra/tiros/blob/main/js/web-vitals.iife.js) to the page.
+
+We got the tti-polyfill from [GoogleChromeLabs/tti-polyfill](https://github.com/GoogleChromeLabs/tti-polyfill/blob/master/tti-polyfill.js) (archived in favor of the [First Input Delay](https://web.dev/fid/) metric).
+We got the web-vitals javascript from [GoogleChrome/web-vitals](https://github.com/GoogleChrome/web-vitals) by building it ourselves with `npm run build` and then copying the `web-vitals.iife.js` (`iife` = immediately invoked function execution)
+
+Then we execute the following javascript on that page ([link to source](https://github.com/dennis-tra/tiros/blob/main/js/measurement.js)):
+
+```javascript
```

Comment:
Per above, link to the actual source?

Reply:
I removed the source code bits from here.